File: src/services/blockService.js
const db = require('../config/db');
const queueService = require('./queueService');

exports.createBlock = async (userId, designId, trx = null, referenceId = null) => {
    const execute = async (t) => {
        // 1. Verify User Constraints (Dynamic Limits)
        const user = await t('users').where({ user_id: userId }).first();
        const strikes = user ? (user.strike_count || 0) : 0;
        const maxActiveBlocks = strikes >= 2 ? 1 : (parseInt(process.env.ACTIVE_BLOCK_LIMIT) || 2);

        const activeBlocksCount = await t('blocks')
            .where({ user_id: userId, status: 'active' })
            .count('* as count')
            .first();

        if (activeBlocksCount.count >= maxActiveBlocks) {
            throw new Error(`Active block limit reached (${maxActiveBlocks}). Strike penalty active: ${strikes >= 2}`);
        }

        // 2. Verify Design Availability
        const design = await t('designs')
            .where({ design_id: designId, availability_status: 'available' })
            .first();

        if (!design) {
            throw new Error('Design is currently unavailable or already blocked.');
        }

        // 3. Verify Wallet Balance
        const wallet = await t('wallets').where({ user_id: userId }).first();
        const blockFee = parseInt(process.env.BLOCK_FEE_MIN); // For MVP, we use the static min fee

        if (!wallet || wallet.balance < blockFee) {
            throw new Error(`Insufficient wallet balance. Minimum block fee required: ₹${blockFee}`);
        }

        // 4. Calculate Expiry (T + 48h)
        const expiryTime = new Date();
        expiryTime.setHours(expiryTime.getHours() + parseInt(process.env.BLOCK_EXPIRY_HOURS));

        // 5. Execute Block Creation
        const designPrice = (design.weight * design.gold_rate_snapshot) + design.making_charge_snapshot;

        const [blockId] = await t('blocks').insert({
            user_id: userId,
            design_id: designId,
            price_snapshot: designPrice,
            status: 'active',
            expiry_time: expiryTime
        }, ['block_id']);

        // 6. Lock Design
        await t('designs')
            .where({ design_id: designId })
            .update({ availability_status: 'blocked' });

        // 7. Ledger Entry (Debit)
        await t('wallet_transactions').insert({
            wallet_id: wallet.wallet_id,
            amount: -blockFee,
            type: 'debit',
            source_type: 'block_payment',
            description: `Commitment fee for design #${designId}`,
            reference_id: referenceId // Added for security audit
        });

        // 8. Update Wallet Balance
        await t('wallets')
            .where({ user_id: userId })
            .update({
                balance: wallet.balance - blockFee,
                last_updated: t.fn.now()
            });

        // 9. Emit Notification Event
        await queueService.publish('block.created', {
            user_id: userId,
            design_id: designId,
            expiry_time: expiryTime
        }, t);

        return {
            block_id: blockId,
            expiry_time: expiryTime,
            locked_price: designPrice
        };
    };

    if (trx) {
        return await execute(trx);
    } else {
        return await db.transaction(execute);
    }
};

/**
 * Service to handle Block Expiry (Called by worker)
 */
exports.expireBlock = async (blockId) => {
    return await db.transaction(async (trx) => {
        const block = await trx('blocks').where({ block_id: blockId, status: 'active' }).first();
        if (!block) return;

        // 1. Release Design
        await trx('designs')
            .where({ design_id: block.design_id })
            .update({ availability_status: 'available' });

        // 2. Update Block Status
        await trx('blocks')
            .where({ block_id: blockId })
            .update({ status: 'expired' });

        // 3. Cancel Orphaned Appointments
        await trx('appointments')
            .where({ block_id: blockId, status: 'requested' })
            .update({ status: 'cancelled' });

        // 4. Emit Notification Event
        await queueService.publish('block.expired', {
            user_id: block.user_id,
            design_id: block.design_id
        }, trx);

        // Note: Intent Engine philosophy - Balance is NOT returned on expiry.
        // It remains in the system for platform decay or admin correction.
    });
};

File: src/services/block_service.py
import os
from datetime import datetime, timedelta
from src.config.db import get_db
from src.services.queue_service import queue_service

def create_block(user_id, design_id, db_conn=None, reference_id=None):
    def execute(db):
        user = db.execute('SELECT * FROM users WHERE user_id = ?', (user_id,)).fetchone()
        strikes = user.get('strike_count', 0) if user else 0
        limit_env = os.environ.get('ACTIVE_BLOCK_LIMIT', '2')
        max_active_blocks = 1 if strikes >= 2 else int(limit_env)

        active_blocks = db.execute(
            "SELECT COUNT(*) as count FROM blocks WHERE user_id = ? AND status = 'active'", 
            (user_id,)
        ).fetchone()

        if active_blocks and active_blocks['count'] >= max_active_blocks:
            raise ValueError(f"Active block limit reached ({max_active_blocks}). Strike penalty active: {strikes >= 2}")

        design = db.execute(
            "SELECT * FROM designs WHERE design_id = ? AND availability_status = 'available'",
            (design_id,)
        ).fetchone()

        if not design:
            raise ValueError('Design is currently unavailable or already blocked.')

        wallet = db.execute('SELECT * FROM wallets WHERE user_id = ?', (user_id,)).fetchone()
        block_fee = int(os.environ.get('BLOCK_FEE_MIN', '100'))

        if not wallet or wallet['balance'] < block_fee:
            raise ValueError(f"Insufficient wallet balance. Minimum block fee required: ₹{block_fee}")

        expiry_hours = int(os.environ.get('BLOCK_EXPIRY_HOURS', '48'))
        expiry_time = (datetime.utcnow() + timedelta(hours=expiry_hours)).strftime('%Y-%m-%d %H:%M:%S')

        design_price = (design['weight'] * design['gold_rate_snapshot']) + design['making_charge_snapshot']

        # 5. Execute Block Creation
        cursor = db.execute('''
            INSERT INTO blocks (user_id, design_id, price_snapshot, status, expiry_time)
            VALUES (?, ?, ?, ?, ?)
        ''', (user_id, design_id, design_price, 'active', expiry_time))
        block_id = cursor.lastrowid

        # 6. Lock Design
        db.execute("UPDATE designs SET availability_status = 'blocked' WHERE design_id = ?", (design_id,))

        # 7. Ledger Entry (Debit)
        db.execute('''
            INSERT INTO wallet_transactions (wallet_id, amount, type, source_type, description, reference_id)
            VALUES (?, ?, ?, ?, ?, ?)
        ''', (wallet['wallet_id'], -block_fee, 'debit', 'block_payment', f"Commitment fee for design #{design_id}", reference_id))

        # 8. Update Wallet Balance
        now = datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')
        db.execute('''
            UPDATE wallets SET balance = balance - ?, last_updated = ? WHERE user_id = ?
        ''', (block_fee, now, user_id))

        # 9. Emit Notification Event
        queue_service.publish('block.created', {
            'user_id': user_id,
            'design_id': design_id,
            'expiry_time': expiry_time
        }, db)

        return {
            'block_id': block_id,
            'expiry_time': expiry_time,
            'locked_price': design_price
        }

    if db_conn:
        return execute(db_conn)
    else:
        with get_db() as db:
            return execute(db)

def expire_block(block_id):
    with get_db() as db:
        block = db.execute("SELECT * FROM blocks WHERE block_id = ? AND status = 'active'", (block_id,)).fetchone()
        if not block:
            return

        db.execute("UPDATE designs SET availability_status = 'available' WHERE design_id = ?", (block['design_id'],))
        db.execute("UPDATE blocks SET status = 'expired' WHERE block_id = ?", (block_id,))
        db.execute("UPDATE appointments SET status = 'cancelled' WHERE block_id = ? AND status = 'requested'", (block_id,))

        queue_service.publish('block.expired', {
            'user_id': block['user_id'],
            'design_id': block['design_id']
        }, db)
        
        # Balance is NOT returned on expiry.
